### PR TITLE
planner: fix cte-schema-clone will clone the old hashcode of its column if any

### LINKED ISSUE #35404: TiDB v6.1: ERROR 1105 (HY000): Can't find column test.tbl.id in schema Column: [Column#7] Unique key: [[test.tbl.id]]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table if exists tbl;
create table tbl (id int);
with t1 as (select id from tbl), t2 as (select a.id from t1 a join t1 b on a.id = b.id) select * from t2 where id in (select id from t2);
```

### 2. What did you expect to see? (Required)

return success with empty set.

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): Can't find column test.tbl.id in schema Column: [Column#7] Unique key: [[test.tbl.id]]
```

### 4. What is your TiDB version? (Required)

SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v6.1.0
Edition: Community
Git Commit Hash: 1a89decdb192cbdce6a7b0020d71128bc964d30f
Git Branch: heads/refs/tags/v6.1.0
UTC Build Time: 2022-06-05 05:15:11
GoVersion: go1.18.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.001 sec)

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close #35404

Problem Summary:
After #33158, the predicate pushdown for CTE has been supported. While some cases will have some problems when it comes to projection elimination using column's hashcode mapping.

In building dataSource,  we clone CTE schema with some newly assigned uniqueID when referring to a CTE. In PR #33158, code as following
```
for i, col := range lp.schema.Columns {
	lp.cte.ColumnMap[string(col.HashCode(nil))] = prevSchema.Columns[i]
}
```
here will generate hashcode in columns of the schema of CTE when building CTE itself.
Later when building the main select clause, CTE reference will clone CTE's schema and assign new uniqueIDs, here should clean these hashcodes because they are not valid anymore.(new uniqueID)

### What is changed and how it works?
Just a quick test for verity resolveIndices error when select statement with join of CTE reference

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
planner: fix cte-schema-clone will clone the old hashcode of its column if any
```

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -6889,6 +6893,8 @@ func getResultCTESchema(seedSchema *expression.Schema, svar *variable.SessionVar
 		col.RetType = col.RetType.Clone()
 		col.UniqueID = svar.AllocPlanColumnID()
 		col.RetType.DelFlag(mysql.NotNullFlag)
+		// Since you have reallocated unique id here, the old-cloned-cached hash code is not valid anymore.
+		col.CleanHashCode()
```
Comment (chrysan) [inline planner/core/logical_plan_builder.go:6897]: getResultCTESchema is called in 3 places, can you add unit tests to cover these paths?

Comment (AilinKid) [reply planner/core/logical_plan_builder.go:6893]: yes, it's ok. just a cache clean, hashcode still can be accessed with recomputation when using

File: expression/column.go [hunk 2]
```diff
@@ -490,6 +490,15 @@ func (col *Column) HashCode(_ *stmtctx.StatementContext) []byte {
 	return col.hashcode
 }
 
+// CleanHashCode will clean the hashcode you may be cached before. It's used especially in schema-cloned & reallocated-uniqueID's cases.
+func (col *Column) CleanHashCode() {
+	if len(col.hashcode) != 0 {
+		col.hashcode = col.hashcode[:0]
+	} else {
+		col.hashcode = make([]byte, 0, 9)
```
Comment (winoros) [inline expression/column.go:498]: I think we can just use this code branch.

the `[:0]` won't clean the memory usage. The two different unique id might refer to the same memory.

Comment (AilinKid) [reply expression/column.go:498]: make sense

**Pull request merged**